client/task_runner.go

```diff
@@ -587,29 +587,37 @@ OUTER:
 }
 
 // deriveVaultToken derives the Vault token using exponential backoffs. It
-// returns the Vault token and whether the token is valid. If it is not valid we
-// are shutting down
-func (r *TaskRunner) deriveVaultToken() (string, bool) {
+// returns the Vault token and whether the manager should exit.
+func (r *TaskRunner) deriveVaultToken() (token string, exit bool) {
 	attempts := 0
 	for {
 		tokens, err := r.vaultClient.DeriveToken(r.alloc, []string{r.task.Name})
 		if err == nil {
-			return tokens[r.task.Name], true
+			return tokens[r.task.Name], false
+		}
+
+		// Check if we can't recover from the error
+		if rerr, ok := err.(*structs.RecoverableError); !ok || !rerr.Recoverable {
```

Comment:
[nit] Maybe put this logic in an `IsRecoverableError(error)` helper function since it's duplicated in a couple different forms?